scripts/dts: Walk chained interrupt controllers

galak opened the pull request:
Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
scripts/dts: Generate IRQ values for multi-level IRQs

When we have multi-level (ie chained interrupt controllers) Zephyr has a
schemee to encode the multi-level and IRQ values along the level's into
a single 32-bit value.  This is the "IRQ" value expected by Zephyr APIs.

The encoding scheme is documented here:

doc/reference/kernel/other/interrupts.rst

Update the device tree generation to walk the interrupt levels and
generate the expected encoded value for the IRQ.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -440,7 +440,14 @@ def irq_name_alias(irq, cell_name):
             ident = "IRQ_{}".format(irq_i)
-            if cell_name != "irq":
+            if cell_name == "irq":
+                # TODO: ADD COMMENT ABOUT ZEPHYR IRQ NUM
```
Yeah, some more comments would be nice here, to explain how `irq` is special, why the values are concatenated, etc. Should document the special `irq` behavior in `binding-template.yaml` as well (or in the `.rst`).

Does the code make any assumptions, like all interrupts being sent to the same place? Would be nice to mention those as well.

galak replied:
Not sure what to document in `binding-template.yaml`, before this change we should document how multi-level IRQs are "hacked" in.  This fixes it so its more natural and reflects the hardware in the DTS (and remove's the Zephyr specific hack/encoding).

ulfalizer replied:
Maybe it could be a link to that interrupts page on https://docs.zephyrproject.org/latest/guides/dts/index.html instead, if it's documented there.

Doesn't have to go in this PR though.

galak replied:
that make's sense, to have a section there about IRQ define generation.  Opened issue #18137 for this.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -440,7 +440,14 @@ def irq_name_alias(irq, cell_name):
+                irq_ctrl = irq.controller
+                while irq_ctrl.interrupts:
+                    cell_value <<= 8
+                    cell_value |= irq_ctrl.interrupts[0].specifier['irq']
```
Nit: Double quotes for consistency.

Should add error checking and not assume that the "irq" key exists. Maybe some kind of helper function for the `cell_name == "irq"` case would work.

galak replied:
Fixed the quote's.  Added a helper function.

What were you thinking w/regards to the error checking?

ulfalizer replied:
Just that "irq" isn't known to be in there (if someone messed up the bindings). Should fail with something nicer than a `KeyError: 'irq'` from Python in that case.

galak committed:
irq: rv32m1: Fixup IRQ values for multi-level IRQ handling

Remove the handcoded multi-level IRQ values in device tree.  We now are
able to generate the encoded multi-level IRQ value.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
irq: intel_s1000: Fixup IRQ values for multi-level IRQ handling

Remove the handcoded multi-level IRQ values in device tree.  We now are
able to generate the encoded multi-level IRQ value.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

ulfalizer commented on dts/bindings/interrupt-controller/openisa,rv32m1-intmux-ch.yaml:
```diff
@@ -0,0 +1,30 @@
+
+#
+# SPDX-License-Identifier: Apache-2.0
+#
```
Could shave three lines with

```
# Copyright (c) 2019, Linaro Limited
#
# SPDX-License-Identifier: Apache-2.0
```

galak replied:
Pulled the base patch from PR #18135

ulfalizer replied:
Could it be fixed there?

galak replied:
> Could it be fixed there?

I think I did.

ulfalizer commented on dts/riscv/rv32m1.dtsi:
```diff
@@ -94,24 +94,188 @@
+				reg = <0x1c0 0x40>;
+				label = "INTMUX0_CH7";
+				status = "disabled";
+			};
```
Maybe a macro that takes an index and an address (for `reg`) could be used here. Lots of typo-prone duplication.

galak replied:
Pulled the base patch from PR #18135

ulfalizer replied:
Noticed it gets messy with the C preprocessor because of the hash marks and stuff anyway, so never mind.

ulfalizer commented on dts/riscv/rv32m1.dtsi:
```diff
@@ -94,24 +94,188 @@
+				reg = <0x1c0 0x40>;
+				label = "INTMUX1_CH7";
+				status = "disabled";
+			};
```
Ditto here.

galak replied:
Pulled the base patch from PR #18135

ulfalizer replied:
Noticed it gets messy with the C preprocessor because of the hash marks and stuff anyway, so never mind.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -435,11 +435,23 @@ def irq_name_alias(irq, cell_name):
+        irq_ctrl = irq.controller
+        while irq_ctrl.interrupts:
+            irq_num = (irq_num + 1) << 8
+            irq_num |= irq_ctrl.interrupts[0].specifier["irq"]
```
Should check that the "irq" key appears in `irq_ctrl.interrupts[0].specifier`, and give a helpful error otherwise.

The goal is to never fail with a generic Python error in the new code.

galak replied:
fixed.

ulfalizer replied:
Could give the path to the binding (`irq_ctrl.binding_path`) instead/as well, since that's where it's missing.

```python
_err("expected 'irq' to appear in '#cells' in {} (for {!r})"
     .format(irq_ctrl.binding_path, irq_ctrl)) 
```

(Should've replied here instead.)

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -435,11 +435,23 @@ def irq_name_alias(irq, cell_name):
 
+    def encode_zephyr_multi_level_irq(irq, irq_num):
+        # See doc/reference/kernel/other/interrupts.rst for details
+        # on how this encoding works
```
Blank line after doc comment, to just be consistent with elsewhere.

Does this code make any simplifying assumptions? For example, if some interrupt controller in turn sends interrupts to five other controllers (via `interrupts-extended`), then does the code make sense?

Should mention simplifying assumptions in a comment, if there are any.

galak replied:
Can't think of any assumptions.  Should work fine w/regards to `interrupt-extended`  or `interrupt-maps`.

ulfalizer replied:
Okay, fine then.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -435,11 +435,23 @@ def irq_name_alias(irq, cell_name):
             ident = "IRQ_{}".format(irq_i)
-            if cell_name != "irq":
+            if cell_name == "irq":
+                cell_value = encode_zephyr_multi_level_irq(irq, cell_value)
```
Bit icky that this makes the name `cell_value` a lie, but not horrible here...

galak replied:
Yeah, didn't like it but felt adding 'if' bellow or some other special handling of "irq" seemed worse.

dcpleung reviewed (approved):
Looks good on the S1000 changes.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -435,11 +435,27 @@ def irq_name_alias(irq, cell_name):
+        while irq_ctrl.interrupts:
+            irq_num = (irq_num + 1) << 8
+            if "irq" not in irq_ctrl.interrupts[0].specifier:
+                _err("missing 'irq' specifier on {!r}".format(irq_ctrl))
```
Could give the path to the binding (`irq_ctrl.binding_path`) instead/as well, since that's where it's missing.

```python
_err("expected 'irq' to appear in '#cells' in {} (for {!r})"
     .format(irq_ctrl.binding_path, irq_ctrl)) 
```

galak replied:
Aren't we going to get the yaml from the `{!r}`

Does this feel redundant?:

```
expected 'irq' to appear in '#cells' in /home/galak/git/zephyr/dts/bindings/interrupt-controller/openisa,rv32m1-intmux-ch.yaml (for <Device /soc/intmux@4004f000/interrupt-controller@40 in 'rv32m1_vega_ri5cy.dts.pre.tmp', binding /home/galak/git/zephyr/dts/bindings/interrupt-controller/openisa,rv32m1-intmux-ch.yaml>)
```

ulfalizer replied:
Forgot that it was already mentioned there.

Maybe it could be

```python
_err("expected binding for {!r} to have 'irq' in '#cells'"
     .format(irq_ctrl))
```

galak replied:
changed to that.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -435,11 +435,28 @@ def irq_name_alias(irq, cell_name):
+        return irq_num
+
     for irq_i, irq in enumerate(dev.interrupts):
         # We ignore the controller for now
```
Outdated comment?

galak replied:
Removed.

ulfalizer commented on scripts/dts/gen_defines.py:
```diff
@@ -435,11 +435,28 @@ def irq_name_alias(irq, cell_name):
+        while irq_ctrl.interrupts:
+            irq_num = (irq_num + 1) << 8
+            if "irq" not in irq_ctrl.interrupts[0].specifier:
+                _err("Expected binding for {!r} to have 'irq' "
```
Should be `err()` rather than `_err()`, as it's not in edtlib.py.

galak replied:
Changed.

galak committed:
interrupt: Convert RISC-V plic to use multi-level irq support

Utilize the multi-level irq infrastructure and replace custom handling
for PLIC on riscv-privilege SoCs.  The old code offset IRQs in drivers
and various places with RISCV_MAX_GENERIC_IRQ.  Instead utilize Zephyr's
encoded IRQ and replace offsets in drivers with the IRQ define from DTS.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

galak committed:
interrupt_controller: plic: remove plic.h

Fold plic.h into plic.c as the .h only defined two things that are
only used in plic.c.

Signed-off-by: Kumar Gala <kumar.gala@linaro.org>

mbolivar commented on soc/riscv/riscv-privilege/common/soc_common_irq.c:
```diff
@@ -11,12 +11,30 @@
 #include <irq.h>
 
+/**
+ * @brief Get an IRQ's level
```
I think it'd be nice to have generic helpers in an `<interrupt_controller/multilevel.h>` or something in case we need to change the encoding later, but it's not a big deal.

(The current encoding can be pretty wasteful of space if there are IRQ chips with widely varying numbers of IRQs, so I don't expect it to last forever.)

galak replied:
Yeah, felt like that would be a post cleanup after this goes in.

mbolivar reviewed (approved):
I haven't checked everything in detail but the approach and patches seem sane to me.

galak commented:
Depends on #18135

nvlsianpu commented:
^^look like a github web bug - I unsigned only myself, but in result bunch of people were unsigned as well.
